src/core/reactor.cc

```diff
@@ -1921,6 +1924,114 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
     });
 }
 
+future<std::tuple<pid_t, int, int>>
+reactor::spawn(std::string_view pathname,
+               const std::vector<std::string_view>& argv,
+               const std::vector<std::string_view>& env) {
+    return do_with(pid_t{}, std::array<int, 2>{}, std::array<int, 2>{},
+                   sstring(pathname),
+                   posix_spawn_file_actions_t{},
+                   posix_spawnattr_t{},
+                   std::vector<sstring>(argv.begin(),
+                                        argv.end()),
+                   std::vector<sstring>(env.begin(),
+                                        env.end()),
+                   [this](auto& child_pid, auto& cout_pipe, auto& cerr_pipe, auto& pathname, auto& actions, auto& attr, auto& argv, auto& env) {
+        auto create_cout_pipe = _thread_pool->submit<syscall_result<int>>([&cout_pipe] {
+            return wrap_syscall<int>(::pipe2(cout_pipe.data(), O_NONBLOCK));
+        }).then([] (syscall_result<int> ret) {
+            ret.throw_if_error();
+            return make_ready_future();
+        });
+        auto create_cerr_pipe = _thread_pool->submit<syscall_result<int>>([&cerr_pipe] {
+            return wrap_syscall<int>(::pipe2(cerr_pipe.data(), O_NONBLOCK));
+        }).then([] (syscall_result<int> ret) {
+            ret.throw_if_error();
+            return make_ready_future();
+        });
+        return when_all_succeed(std::move(create_cout_pipe),
```

Comment:
It looks like either pipe would leak if the creation of its peer fails

Reply:
yes. will fix the leak in the next revision.